crates/precompile/src/bn128/arkworks.rs

```diff
@@ -0,0 +1,213 @@
+use super::{FQ2_LEN, FQ_LEN, G1_LEN, SCALAR_LEN};
+use crate::PrecompileError;
+use std::vec::Vec; // TODO: Remove this import, `cargo clippy --no-default-features` needs it to disambiguate
+
+use ark_bn254::{Bn254, Fq, Fq2, Fr, G1Affine, G1Projective, G2Affine};
+use ark_ec::{pairing::Pairing, AffineRepr, CurveGroup};
+use ark_ff::{One, PrimeField, Zero};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+
+/// Reads a single `Fq` field element from the input slice.
+///
+/// Takes a byte slice and attempts to interpret the first 32 bytes as an
+/// elliptic curve field element. Returns an error if the bytes do not form
+/// a valid field element.
+///
+/// # Panics
+///
+/// Panics if the input is not at least 32 bytes long.
+#[inline]
+fn read_fq(input_be: &[u8]) -> Result<Fq, PrecompileError> {
+    // TODO: maybe assert input_be is 32 bytes, then change panic note to _exactly_ 32 bytes?
+
+    let mut input_le = [0u8; FQ_LEN];
+    input_le.copy_from_slice(input_be);
+
+    // Reverse in-place to convert from big-endian to little-endian.
+    input_le.reverse();
+
+    Fq::deserialize_uncompressed(&input_le[..])
+        .map_err(|_| PrecompileError::Bn128FieldPointNotAMember)
+}
+/// Reads a Fq2 (quadratic extension field element) from the input slice.
+///
+/// Parses two consecutive Fq field elements as the real and imaginary parts
+/// of an Fq2 element.
+/// The second component is parsed before the first, ie if a we represent an
+/// element in Fq2 as (x,y) -- `y` is parsed before `x`
+///
+/// # Panics
+///
+/// Panics if the input is not at least 64 bytes long.
+#[inline]
+fn read_fq2(input: &[u8]) -> Result<Fq2, PrecompileError> {
+    let y = read_fq(&input[..FQ_LEN])?;
+    let x = read_fq(&input[FQ_LEN..2 * FQ_LEN])?;
+
+    Ok(Fq2::new(x, y))
+}
+
+/// Creates a new `G1` point from the given `x` and `y` coordinates.
+///
+/// Constructs a point on the G1 curve from its affine coordinates.
+///
+/// Note: The point at infinity which is represented as (0,0) is
+/// handled specifically because `AffineG1` is not capable of
+/// representing such a point.
+/// In particular, when we convert from `AffineG1` to `G1`, the point
+/// will be (0,0,1) instead of (0,1,0)
+#[inline]
+fn new_g1_point(px: Fq, py: Fq) -> Result<G1Affine, PrecompileError> {
+    if px.is_zero() && py.is_zero() {
+        Ok(G1Affine::zero())
+    } else {
+        let point = G1Affine::new_unchecked(px, py);
+        if !point.is_on_curve() || !point.is_in_correct_subgroup_assuming_on_curve() {
+            return Err(PrecompileError::Bn128AffineGFailedToCreate);
+        }
```

Comment:
Another small note `||` are short-circuit operations.

- `a || b` will not check `b` if `a` is true

- `a` in this case is `!point.is_on_curve()` so if the point is not on the curve, meaning it is invalid, then we don't check if it is in the correct subgroup. 

Perhaps it would be clearer to write this as:

```rust
if !point.is_on_curve() {
   return return Err(PrecompileError::Bn128AffineGFailedToCreate);
}
if !point.is_in_correct_subgroup_assuming_on_curve() {
   return Err(PrecompileError::Bn128AffineGFailedToCreate);
}
```

Reply:
This is fine, short-circuit operations are okay,